cranelift/isle/isle/src/overlap.rs

```diff
@@ -0,0 +1,876 @@
+    fn add_edge(&mut self, other: RuleId) {
+        if self.edges[other.0] {
+            return;
+        }
+
+        self.degree += 1;
+        self.edges[other.0] = true;
+    }
+
+    /// Remove an edge between this node and another node.
+    fn remove_edge(&mut self, other: RuleId) {
+        if !self.edges[other.0] {
+            return;
+        }
+
+        self.degree -= 1;
+        self.edges[other.0] = false;
+    }
+}
+
+/// A graph of all the rules in the isle source, with bi-directional edges between rules that are
+/// discovered to have overlap problems.
+struct Errors {
+    /// Edges between rules indicating overlap. As the edges are not directed, the edges are
+    /// normalized by ordering the rule ids.
+    nodes: Vec<Node>,
+}
+
+impl Errors {
+    /// Make a new `Errors` graph for collecting overlap information.
+    fn new(len: usize) -> Self {
+        let mut nodes = Vec::with_capacity(len);
+        nodes.resize_with(len, || Node::new(len));
+        Self { nodes }
+    }
+
+    /// True when there are no edges in the graph.
+    fn is_empty(&self) -> bool {
+        self.nodes.iter().all(|node| node.degree == 0)
+    }
+
+    /// Condense the overlap information down into individual errors.
+    fn report(&mut self, env: &Env) -> Vec<Error> {
+        let mut rules: Vec<usize> = self
+            .nodes
+            .iter()
+            .enumerate()
+            .filter_map(|(id, node)| if node.degree == 0 { None } else { Some(id) })
+            .collect();
+
+        rules.sort_by_cached_key(|id| self.nodes[*id].degree);
+
+        let mut errors = Vec::new();
+
+        let get_info = |id| {
+            let rule = env.get_rule(id);
+
+            let src = env.get_source(rule.pos.file);
+            let span = Span::new_single(rule.pos);
+            (src, span)
+        };
+
+        // Work backwards through the ids to find the nodes with the largest conflict first.
+        for id in rules.into_iter().rev() {
+            let node = self.remove_edges(RuleId(id));
+            if node.degree == 0 {
+                continue;
+            }
+
+            // build the real error
+            let mut rules = vec![get_info(RuleId(id))];
+
+            rules.extend(
+                node.edges
+                    .into_iter()
+                    .enumerate()
+                    .filter_map(|(ix, present)| {
+                        if present {
+                            Some(get_info(RuleId(ix)))
+                        } else {
+                            None
+                        }
+                    }),
+            );
+
+            errors.push(Error::OverlapError {
+                msg: String::from("rules are overlapping"),
+                rules,
+            });
+        }
+
+        errors
+    }
+
+    /// Remove all the edges for this rule in the graph, returning the original `Node` contents for
+    /// further processing.
+    fn remove_edges(&mut self, id: RuleId) -> Node {
```

Comment:
The `Errors::remove_edges` method is only called by `report`. I'd suggest inlining it there, and here's why:

If you change `Errors` to contain a `HashMap`, then you could establish an invariant that every node in the map has non-zero degree. That's useful because `Errors::is_empty` can just call `HashMap::is_empty`, and `Errors::report` doesn't need to filter the set of nodes to find non-empty ones.

That invariant automatically holds initially and after any sequence of calls to `add_edge`. But `remove_edges` breaks that invariant unless you're careful. Still, you don't need the invariant once you start looping over the overlaps in `report`, so if you inline this there then you don't have to be careful about the invariant.